lib/utils.js

```diff
@@ -39,6 +39,21 @@ exports.pick = function (obj, arr) {
   return res;
 };
 
+// just like emitter.once(event, listener), but doesn't
+// add additional listeners if any listeners with the
+// key are already attached
+exports.oncePerKey = function oncePerKey(emitter, event, key, listener) {
+  var listeners = emitter.listeners(event).filter(function (listener) {
+    return listener._key === key;
+  });
+  if (listeners.length) {
+    return;
+  }
+  emitter.once(event, listener);
+  listeners = emitter.listeners(event);
+  listeners[listeners.length - 1]._key = key;
+};
+
```

Comment:
I think this will cause problems. The first set of listeners attached with oncePerKey() will reference the state from the first call to replicate()'s closure. If the second or third call to replicate() uses those callbacks referencing the original closure, then all of these var's will be different from the current replication attempt compared to the ones inside the callback: https://github.com/pouchdb/pouchdb/blob/3699-3824/lib/replicate.js#L150

A different way to solve the same leak might be to have the cleanup callbacks all live on the Replication() class so they cannot reference the replicate() function's closure. The ordinary "if existing callback === new callback" check could be used since the cleanup handlers would now be identical between replicate() calls.

And/or instead of calling replicate() multiple times, backOff() could call a private function that doesn't try to install the cleanup handlers. Example:

```javascript
function replicate() {
   install_cleanup_handlers();
   do_replication();
}

function do_replication() {
    // do it
}

function backOff() {
    // wait
    do_replication();
}
```